pkg/types/v1/config_test.go

```diff
@@ -224,4 +225,225 @@ var _ = Describe("Types", Label("types", "config"), func() {
+				spec.PartTable = constants.GPT
+				spec.ExtraPartitions = v1.PartitionList{
+					&v1.Partition{
+						Size: 0,
+					},
+				}
+				err := spec.Sanitize()
+				Expect(err).ToNot(HaveOccurred())
+			})
+		})
+		Describe("ResetSpec sanitize", func() {
+			var spec v1.ResetSpec
+			BeforeEach(func() {
+				spec = v1.ResetSpec{
+					Active: v1.Image{
+						Source: v1.NewEmptySrc(),
+					},
+				}
+			})
+			It("fails with empty source", func() {
+				err := spec.Sanitize()
+				Expect(err).To(HaveOccurred())
+				Expect(err.Error()).To(ContainSubstring("undefined system source to reset"))
+			})
+			It("fails with missing state partition", func() {
+				spec.Active.Source = v1.NewFileSrc("/tmp")
+				err := spec.Sanitize()
+				Expect(err).To(HaveOccurred())
+				Expect(err.Error()).To(ContainSubstring("undefined state partition"))
+			})
+			It("passes if state and source are ready", func() {
+				spec.Active.Source = v1.NewFileSrc("/tmp")
+				spec.Partitions.State = &v1.Partition{
+					MountPoint: "/tmp",
+				}
+				err := spec.Sanitize()
+				Expect(err).ToNot(HaveOccurred())
+			})
+
+		})
+		Describe("UpgradeSpec sanitize", func() {
+			var spec v1.UpgradeSpec
+			BeforeEach(func() {
+				spec = v1.UpgradeSpec{
+					Active: v1.Image{
+						Source: v1.NewEmptySrc(),
+					},
+					Recovery: v1.Image{
+						Source: v1.NewEmptySrc(),
+					},
+				}
+			})
+			Describe("Active upgrade", func() {
+				It("fails with empty source", func() {
+					err := spec.Sanitize()
+					Expect(err).To(HaveOccurred())
+					Expect(err.Error()).To(ContainSubstring("undefined upgrade source"))
+				})
+				It("fails with missing state partition", func() {
+					spec.Active.Source = v1.NewFileSrc("/tmp")
+					err := spec.Sanitize()
+					Expect(err).To(HaveOccurred())
+					Expect(err.Error()).To(ContainSubstring("undefined state partition"))
+				})
+				It("passes if state and source are ready", func() {
+					spec.Active.Source = v1.NewFileSrc("/tmp")
+					spec.Partitions.State = &v1.Partition{
+						MountPoint: "/tmp",
+					}
+					err := spec.Sanitize()
+					Expect(err).ToNot(HaveOccurred())
+				})
+			})
+			Describe("Recovery upgrade", func() {
+				BeforeEach(func() {
+					spec.RecoveryUpgrade = true
+				})
+				It("fails with empty source", func() {
+					err := spec.Sanitize()
+					Expect(err).To(HaveOccurred())
+					Expect(err.Error()).To(ContainSubstring("undefined upgrade source"))
+				})
+				It("fails with missing recovery partition", func() {
+					spec.Recovery.Source = v1.NewFileSrc("/tmp")
+					err := spec.Sanitize()
+					Expect(err).To(HaveOccurred())
+					Expect(err.Error()).To(ContainSubstring("undefined recovery partition"))
+				})
+				It("passes if state and source are ready", func() {
+					spec.Recovery.Source = v1.NewFileSrc("/tmp")
+					spec.Partitions.Recovery = &v1.Partition{
+						MountPoint: "/tmp",
+					}
+					err := spec.Sanitize()
+					Expect(err).ToNot(HaveOccurred())
+				})
```

Comment:
These seem so similar between the diferent `Describe` cases that mabye they could be written once in the beforeEach, and then the differences changed in a context in the justBeforeEach

Reply:
Indeed butmore importantly to me it shows that we have too many things in common in the sanitize methods that could be extracted into a single function and reused or a generic sanitize created for the common parts!